### DISCUSSION TITLE
Explicit column ordering

### DISCUSSION THREAD

**Discussion by seancolsen** [mathesar-foundation/mathesar#1186] [Maintainer discussions]
@kgodey [said](https://github.com/centerofci/mathesar/issues/1185#issuecomment-1069645342)

> I think it would be ideal for the API to return the column's index as a property and the frontend should order by that. Please make an issue for that; until then, it's fine to order by id.

I'm starting a new discussion here to respond.

**Comment by seancolsen** [comment]
@kgodey Can you help me see the value in passing a separate property in the API response? Unless/until we can use the API to re-order columns, it seems to me that the position of the columns in the response array sufficiently communicates their order.

**Reply by kgodey** [reply]
> it seems to me that the position of the columns in the response array sufficiently communicates their order.

Not if they're not ordered correctly, which is what was happening in #1185. If they are ordered correctly, I don't think that index needs to be a separate property. 

In #1185, you said:

> for implementation, can we assume that ordering by the column's `id` is a reliable-enough way to obtain the expected column order?

If you're already parsing each individual column and ordering by a key inside that, it should be `index`, not `id`. `id` is not reliable as a guarantee of order.

**Reply by seancolsen** [reply]
Hmm. When I was reproducing the bug in #1185 I observed that ordering by `id` _would_ produce the correct order, were we to order by `id`. If columns are always _created_ in the correct order, then I'd assume we could add logic on the back end to order the columns by their ids before sending the API response. 

> If you're already parsing each individual column and ordering by a key inside that

We're not. The front end is displaying the columns in the order in which it receives them.

**Reply by seancolsen** [reply]
Note that #1185 is a backend ticket. When I said "for implementation", I meant for _backend_ implementation.

**Reply by kgodey** [reply]
> When I said "for implementation", I meant for _backend_ implementation.

Ah okay, I misunderstood. In that case, the backend should use a column's `index`, not `id` (the backend has this information). It does not need to expose the index to the frontend.

`id` might work in many or most cases, but it's not guaranteed to work because it's assigned by Django for the purposes of the API. `index` is a PostgreSQL database property and is guaranteed.

**Reply by seancolsen** [reply]
> the backend should use a column's index

Cool, can you clarify that in #1185 then?

**Reply by kgodey** [reply]
Done, I updated the issue description.

**Reply by mathemancer** [reply]
Sorry I was slow on this: 
- Columns don't really have a defined "order" in PostgreSQL, and thus
- Index is _not_ a property of a PostgreSQL table (at least not directly). However,
- `attnum` is a property of a PostgreSQL table. Moreover, 
- when displaying a table most clients will order the columns by `attnum`.
- For weird reasons, ordering columns by their IDs will always be in the same order as by their `attnum`s, but this is for reasons specific to our current implementation, and I wouldn't want to use IDs in the long run.

I suggest we order by `attnum`s and remove indices from things entirely.

**Reply by kgodey** [reply]
Okay, sounds good.
